### PR TITLE
Group 4 Application

### PR DESCRIPTION
**Pull request by 132ikl**
Merging into a new branch based off of approximately when we were done with our advanced classes, since #186 isn't merged

To compile and run webserver, make sure to [install emscripten](https://emscripten.org/docs/getting_started/downloads.html) and set up the environment as appropriate for your platform, then in the `Group-04` folder, run:
```
make
make run
```

and then navigate to http://localhost:8000 in your web browser

### CODE REVIEW

**Comment by AmeryJohnson** [comment]
### Implementation Quality
**No warnings during compilation:** There are no warnings. Good job!
**Proper use of namespaces:** Yes the cse namespace was used correctly.
**Adheres to C++ Standard:** Yes there was nothing that I noticed about it that would have undefined behavior.
**Appropriate Modularity:** There are many helper functions for other functions and everything is quite modular great job.
**Uses modern C++ constructs:** Yes uses range based for loops and modern std functions as well.
**Uses standard library algorithms:** Yes it uses many different STL algorithms and they all seem to be used correctly.
**Proper use of pointer handling:** I didn't notice anything that would cause any problems with the pointers that were used.
**Const functions:** Many functions are marked as const and used correctly.

### Testing and Error Checking
**Unit Testing:** Unit tests are extensive and tests for many essential things and is very complete.
**Edge Cases:** I noticed that it tests for a lot of different types of edge cases great job.
**Asserts:** Many asserts are used especially with the class that was made being CseAssert. It seems that that class helped out quite a bit with the development and asserts seem to be used correctly.
**Error Handling:** For error handling it seems that the asserts do most of the work with that although I noticed when I try to apply a certain style to the text it can give different types of errors or it can also say that a wrong user input style is successfully applied even though there was no style applied since the user input was flawed. Even though it does this, any user input will not completely break the application or anything it just gives weird feedback to wrong user input.

File: Group-04/app/web/index.js [hunk 1]
```diff
@@ -1,14 +1,496 @@
...
+    underlineButton.className = "format-button underline-button";
+    underlineButton.addEventListener("click", () => {
+      toggle_monostate("underline");
+      updateOutputs();
+    });
+    
+    // Color button
+    const colorButton = document.createElement("button");
+    colorButton.textContent = "A";
+    colorButton.className = "format-button color-button";
```
Comment (graceKC04) [inline Group-04/app/web/index.js:198]: User Experience:
The “A color” button is somewhat confusing. It visually suggests it is clickable, as the color changes upon clicking, but it doesn’t perform any action. A clearer approach might be to either remove this button or ensure it opens the color picker, making the interface more intuitive.

File: Group-04/app/web/index.js [hunk 2]
```diff
@@ -1,14 +1,496 @@
...
+    formatButtonsContainer.appendChild(underlineButton);
+    formatButtonsContainer.appendChild(colorPicker);
+    formatButtonsContainer.appendChild(colorButton);
+    
+    // Add spacing after format buttons
+    bar.appendChild(formatButtonsContainer);
+    
+    // Custom format section
+    const customFormatContainer = document.createElement("div");
+    customFormatContainer.className = "custom-format-container";
```
Comment (graceKC04) [inline Group-04/app/web/index.js:255]: User Experience:
 The custom box functionality could be hidden for a cleaner interface. Instead of having it visible at all times, it could be tucked behind a "Custom" button. Additionally, when the app first opened, it wasn’t clear whether clicking "Apply" was necessary for every change or just when using the custom color picker. Having the custom box appear only after selecting the “Custom” button would reduce confusion.

File: Group-04/app/web/index.js [hunk 3]
```diff
@@ -1,14 +1,496 @@
...
+
+    // Left-side input bar
+    const bar = document.getElementById("toolbar");
+    const wrappedInput = MakeInput.textInput("", "Type here...");
+    wrappedInput.style.width = "100%";
+    bar.appendChild(wrappedInput);
+
+    // Create formatting buttons
+    const formatButtonsContainer = document.createElement("div");
+    formatButtonsContainer.className = "format-buttons";
```
Comment (graceKC04) [inline Group-04/app/web/index.js:145]: User Experience:
Having commonly used tools available made it much quicker to edit text, good job!

File: Group-04/app/web/index.js [hunk 4]
```diff
@@ -1,14 +1,496 @@
...
+      // Adjust text color based on background color brightness for readability
+      const r = parseInt(color.slice(1, 3), 16);
+      const g = parseInt(color.slice(3, 5), 16);
+      const b = parseInt(color.slice(5, 7), 16);
+      const brightness = (r * 299 + g * 587 + b * 114) / 1000;
+      colorMessage.style.color = brightness > 125 ? 'black' : 'white';
+      
+      bar.appendChild(colorMessage);
+      
+      // Remove the message after a delay
```
Comment (graceKC04) [inline Group-04/app/web/index.js:235]: User Experience:
Every time the color is dragged, an alert is triggered, which can quickly overwhelm users with a long list of color updates. A potential fix could be to limit these alerts to specific actions, such as confirming the color change or showing alerts only for significant updates.

File: Group-04/app/Bindings.hpp
Comment (DominikLeisinger) [inline Group-04/app/Bindings.hpp:1]: Great code design on this, perfectly organized and very easy to read, plus  concise informative comments.

File: Group-04/app/RichTextState.hpp [hunk 5]
```diff
@@ -0,0 +1,249 @@
...
+#include "SerializerRTF.hpp"
+
+class RichTextState {
+ public:
+  using Text = cse::BasicRichText<char, cse::String>;
+
+  RichTextState() {
+    // HTML (no header/footer)
+    m_output_map["html"] = {"HTML", cse::SerializerHTML<char, cse::String>()};
+    m_output_map["html"].second.header = std::nullopt;
```
Comment (DominikLeisinger) [inline Group-04/app/RichTextState.hpp:29]: good use of std::nullopt here

**Review by graceKC04 — COMMENTED** [review]
# User Experience Review (Apologies for the length)

Overall, the application offers a solid user experience but could benefit from a few adjustments to improve usability.  
(-) indicates a potential improvement, and (+) indicates an appreciated feature.

## Intuitiveness & Navigation
The app is easy to use, with most features logically placed.

- (-) Every time the color is dragged, an alert is triggered, which can quickly overwhelm users with a long list of color updates. A potential fix could be to limit these alerts to specific actions, such as confirming the color change or showing alerts only for significant updates.
- (+) The undo/redo feature is well implemented and provides a useful safety net for users who might make mistakes. This is a thoughtful inclusion.
- (+) The ability to highlight text for editing is intuitive and straightforward, making the process even more user-friendly.

## Confusion & Clarity
The app has a few areas where clarity could be improved, but overall, it does not hinder the ability to utilize key features.

- (-) The “A color” button is somewhat confusing. It visually suggests it is clickable, as the color changes upon clicking, but it doesn’t perform any action. A clearer approach might be to either remove this button or ensure it opens the color picker, making the interface more intuitive.
- (-) The custom box functionality could be hidden for a cleaner interface. Instead of having it visible at all times, it could be tucked behind a "Custom" button. Additionally, when the app first opened, it wasn’t clear whether clicking "Apply" was necessary for every change or just when using the custom color picker. Having the custom box appear only after selecting the “Custom” button would reduce confusion.
- (+) Having commonly used tools available made it much quicker to edit text.

## Bug

- (-) Multiple color changes occasionally break some text. A significant bug occurs when attempting to change colors. Specifically, after setting “Hello” to orange and “world!” to blue, and then trying to change both to black, the colors fail to update. Additionally, the text becomes uneditable. This issue disrupts the app's flow and needs to be addressed. A possible fix could be ensuring that color changes reset correctly and that text remains editable.
- (-) Custom not updating in the HTML file type view (as seen in demo).

## Possible Features

- **Cursor Indicator**: Adding a cursor or some form of indicator to highlight the current area being edited would be beneficial. When trying to insert new text between existing words, it can be difficult to see where you're editing unless a portion of the text is already selected. A subtle indicator could help users know exactly where their edits will appear.

- **New Line Functionality for Large Texts**: A feature that allows users to work with large blocks of text more effectively would be valuable. Currently, it may be difficult to edit long texts without being able to insert line breaks. Adding a way to create and navigate new lines would make editing large amounts of text much more manageable.

- **Download and Upload Text in Various File Formats**: Providing the ability to download text in a variety of formats (HTML, LaTeX, Markdown, and RTF) would greatly enhance flexibility. This feature would allow users to easily use their texts in other spaces. (I know this was briefly discussed in the group presentation.)

## Overview

The application is overall useful and intuitive, with a few areas for improvement. The most pressing issue is the bug related to color changes, but otherwise, the interface functions well.

# Code Design Review

Amazing job on clear commentating and consistent formatting. Honestly, I was really impressed with how clear everything was. I think with how well everything is documented, future use and development of your code will be fairly straightforward.

File: Group-04/app/RichTextState.hpp [hunk 6]
```diff
@@ -0,0 +1,249 @@
...
+    save_state();
+
+    cse::TextFormat fmt{id, data};
+    if (m_edit.second > m_edit.first)
+      m_text.apply_format(fmt, m_edit.first, m_edit.second);
+    m_format_map[id] = data;
+  }
+
+  void set_monostate(cse::TextFormat::FormatID id) {
+    set_format(id, std::monostate());
```
Comment (DominikLeisinger) [inline Group-04/app/RichTextState.hpp:105]: another modern c++ construct, great! maybe a small function comment for the reasoning, as well as the meaning for a newbie?

File: Group-04/app/web/index.css
Comment (DominikLeisinger) [inline Group-04/app/web/index.css:1]: Even with minimal css experience, your comments and organized code made it simple for me to understand what was happening throughout this file, great job!

File: Group-04/app/web/index.js
Comment (DominikLeisinger) [inline Group-04/app/web/index.js:1]: Code Design was fantastic here, all relevant code was grouped together, all with brief informative comments and no uneccessary code

File: Group-04/lib/BitVector.hpp [hunk 7]
```diff
@@ -21,30 +21,30 @@ concept BoolGenerator = (std::invocable<T, size_t> &&
 // BitVectors of varying sizes.
 class BitVector {
  private:
-  // Element type
+  // Element type (must be unsigned and integral)
   using bv_type = uint64_t;
   // A std::byte filled with ones
-  static constexpr bv_type ALL_ONE = bv_type(0) - bv_type(1);
+  static constexpr bv_type ALL_ONE = std::numeric_limits<bv_type>::max();
```
Comment (DominikLeisinger) [inline Group-04/lib/BitVector.hpp:27]: Great uses of constxpr here instead of magic numbers!

File: Group-04/lib/BitVector.hpp
Comment (DominikLeisinger) [inline Group-04/lib/BitVector.hpp:1]: This entire file demonstrates great knowledge of advanced C++ concepts, all along with comments to help me understand whats going on the whole way. Love the use of the nested proxy class to make the code more concise and less complex.

File: Group-04/lib/CseAssert.hpp
Comment (DominikLeisinger) [inline Group-04/lib/CseAssert.hpp:1]: Extremely well written code for these custom asserts, love the use of std::nullopt and std::equality_comparable_with. Again my only point would be to add some informative comments, some of the functions after the BinaryAssertArgs class took some time to understand

File: Group-04/lib/CseString.hpp
Comment (DominikLeisinger) [inline Group-04/lib/CseString.hpp:1]: Love this file, great use of your custom asserts for the audited string and all functions had very detailed Doxy comments.

File: Group-04/lib/RichText.hpp
Comment (DominikLeisinger) [inline Group-04/lib/RichText.hpp:1]: Use of std::function, std::variant, std::derived_from, std::holds_alternative, and ranges along with the use of [[attributes]] all show an in-depth understanding of C++ much farther than this class has gone. And Doxy comments, amazing job.

**Comment by nigamsa1** [comment]
**Testing and Error Checking:**

Tests are organized very well and are easy to follow.

Could check what calling undo() with no previous moves does, or redo() when nothing is to redo

Initial state tests assume non‐empty text. Consider adding more tests for empty strings

You test that Markdown output contains the text, but don’t check that the actual Markdown syntax is correct.

Overall, your tests are very extensive and thorough and I do not have much other than small nitpicks to say! Good job!

**Comment by DominikLeisinger** [comment]
**Code Design**:
Overall extremely well written, all very organized and grouped with other relevant code. As someone with little to no HTML, LaTeX, CSS, or JS knowledge your web folder files were all relatively easy to grasp thanks to your clearly separated and commented code. There clearly was a lot of thought that went into your project, and it shows with just the sheer amount of code that went into this. You avoided magic numbers with constexpr, the interface was very intuitive, most of the variable identifiers were informative immediately. My only nitpick in the entire project was inconsistent commenting. You had several different languages which obviously changes how the comments are going to look, but your entire web folder was commented concisely and often. (No issues there) For the C++ files, there was some differences in how you decided to make comments, some using standard Doxy format, some just a '//' line, and some lacked comments where I thought they may be needed( I marked this where observed, CseAssert as an example). Other than that, this completely blew me away. Exceeded expectations.

**Use of Modern C++**:
I made more than several comments detailing how impressed I was with the number of C++ constructs that were unknown to me before today. std::equality_comparable_with, std::variant, std::holds_alternative and many more. I love this project specifically because it was a great learning experience. You all used many detailed templates, [[attributes]] for functions, and a custom Assert file that was very well written and used many times throughout other files. I have nothing but good things to say about this section, again, great job.

File: Group-04/app/web/index.js (hunk 1 above)
Comment (nigamsa1) [reply Group-04/app/web/index.js:198]: I agree!

File: Group-04/app/RichTextState.hpp [hunk 8]
```diff
@@ -0,0 +1,249 @@
...
+  void apply_header(int lvl, size_t start, size_t end) {
+    m_text.apply_format(cse::TextFormat("header", lvl), start, end);
+  }
+
+ private:
+  Text m_text{"Hello World!"};
+  std::map<cse::TextFormat::FormatID, cse::TextFormat::FormatData> m_format_map;
+  std::map<std::string, std::pair<std::string, Text::Serializer>> m_output_map;
+  std::pair<size_t, size_t> m_edit{0, 0};
+  std::vector<Text> undo_stack, redo_stack;
```
Comment (caden-james) [inline Group-04/app/RichTextState.hpp:243]: Since you're just using this like a stack, wouldn't a deque make more sense? It would be faster for larger storage sizes.

File: Group-04/app/RichTextState.hpp [hunk 9]
```diff
@@ -0,0 +1,249 @@
...
+    m_text.apply_format(cse::TextFormat("link", u), start, end);
+  }
+  void apply_header(int lvl, size_t start, size_t end) {
+    m_text.apply_format(cse::TextFormat("header", lvl), start, end);
+  }
+
+ private:
+  Text m_text{"Hello World!"};
+  std::map<cse::TextFormat::FormatID, cse::TextFormat::FormatData> m_format_map;
+  std::map<std::string, std::pair<std::string, Text::Serializer>> m_output_map;
```
Comment (caden-james) [inline Group-04/app/RichTextState.hpp:241]: Does order matter? If not I feel like an unordered_map would just be faster.

File: Group-04/app/RichTextState.hpp [hunk 10]
```diff
@@ -0,0 +1,249 @@
...
+  // Single‑arg convenience formatting
+  void apply_bold() { set_format("bold", std::monostate{}); }
+  void apply_italic() { set_format("italic", std::monostate{}); }
+  void apply_underline() { set_format("underline", std::monostate{}); }
+  void apply_strikethrough() { set_format("strikethrough", std::monostate{}); }
+  void apply_color(const std::string& c) { set_format("color", c); }
+  void apply_link(const std::string& u) { set_format("link", u); }
+  void apply_header(int lvl) { set_format("header", lvl); }
+
+  // Range‑based overloads (the tests call these with start,end)
```
Comment (caden-james) [inline Group-04/app/RichTextState.hpp:212]: For this next section, I feel like sound boundary checks for the start and end variables would be avoiding potential issues in the future.

File: Group-04/lib/BitVector.hpp [hunk 11]
```diff
@@ -21,30 +21,30 @@ concept BoolGenerator = (std::invocable<T, size_t> &&
...
   // updating the m_num_set
-  void truncate_element(size_t byte, size_t keep);
+  void truncate_element(size_t el, size_t keep);
 
   // Manages the case in pattern set where
-  // only one byte needs changing
+  // only one element needs changing
   BitVector& pattern_set_one(size_t start, size_t count, bv_type pattern);
 
   // Convert a number of bits
```
Comment (caden-james) [inline Group-04/lib/BitVector.hpp:50]: Really cool use of bit manipulation here. Can't leave a comment directly but I feel like the destructor could be [[nodiscard]] right?

File: Group-04/lib/BitVector.hpp [hunk 12]
```diff
@@ -21,30 +21,30 @@ concept BoolGenerator = (std::invocable<T, size_t> &&
...
-  // Keep the bottom "keep" bits of the given byte, discarding the rest and
+  // Keep the bottom "keep" bits of the given element, discarding the rest and
   // updating the m_num_set
-  void truncate_element(size_t byte, size_t keep);
+  void truncate_element(size_t el, size_t keep);
 
   // Manages the case in pattern set where
-  // only one byte needs changing
+  // only one element needs changing
   BitVector& pattern_set_one(size_t start, size_t count, bv_type pattern);
```
Comment (caden-james) [inline Group-04/lib/BitVector.hpp:48]: I think you should have a check to make sure the pattern isn't 0 when count > 0.

File: Group-04/lib/BitVector.hpp (hunk 11 above)
Comment (caden-james) [inline Group-04/lib/BitVector.hpp:50]: Also, for the resize operation, you could use some of the SSO optimizations we learned for bitvectors.

File: Group-04/lib/RichText.hpp [hunk 13]
```diff
@@ -400,6 +402,12 @@ class BasicRichText {
    * @param end End (exclusive) of range to apply format to
    */
   void apply_format(const TextFormat& format, const IndexSet& indices) {
```
Comment (caden-james) [inline Group-04/lib/RichText.hpp:404]: You should have a reserve() for the output string considering you have a known size estimate.

File: Group-04/lib/RichText.hpp [hunk 14]
```diff
@@ -598,11 +702,13 @@ class BasicRichText {
...
+            reapply_rule_iter->rule.StartToken(reapply_rule_iter->format);
+        if (all_deactivated) {
           all_deactivated = false;
-        else
+          tracker_iter = trackers.begin();
+        } else {
           ++tracker_iter;
+        }
       } while (tracker_iter != trackers.end());
     }
```
Comment (caden-james) [inline Group-04/lib/RichText.hpp:714]: For the erase function, I would retrun the iterator for the next valid position for better STL compatibility.

File: Group-04/app/Bindings.hpp [hunk 15]
```diff
@@ -0,0 +1,108 @@
...
+
+#include <cstddef>
+#include <string>
+#include <vector>
+
+#include "RichTextState.hpp"
+
+using namespace emscripten;
+
+EMSCRIPTEN_BINDINGS(rich_text_module) {
```
Comment (caden-james) [inline Group-04/app/Bindings.hpp:18]: I would also include bindings for the move/assignment constructor just to be more efficient.

File: Group-04/lib/CseAssert.hpp [hunk 16]
```diff
@@ -80,6 +84,11 @@ void assert_fail() { throw AssertTestException(); }
...
 #endif
 
+#elifdef __EMSCRIPTEN__
+[[noreturn]] void assert_fail() {
+  EM_ASM({alert("Assertion failed (see console)")});
+  std::abort();
+}
 #else
 [[noreturn]] void assert_fail() { std::abort(); }
 #endif
```
Comment (caden-james) [inline Group-04/lib/CseAssert.hpp:94]: Love the printable concept used here but it should constrain to basic_ostream just to be more precise with the output.

File: Group-04/lib/CseString.hpp
Comment (caden-james) [inline Group-04/lib/CseString.hpp:1]: Love a lot of the modern C++ in this file but I would just add noexcept on the move operations just to be thorough.

File: Group-04/lib/SerializerHTML.hpp
Comment (caden-james) [inline Group-04/lib/SerializerHTML.hpp:1]: This class is also great, really not many problems, but the SERIALIZER_HTML_HEADER should be constexpr and I think string_view would be used there to be more consistent.

File: Group-04/lib/SerializerHTML.hpp [hunk 17]
```diff
@@ -33,6 +33,12 @@ auto SerializerHTML() {
...
 
+  out.AddRule("underline", str_to_underlying<CharT, Underlying>("<u>"),
+              str_to_underlying<CharT, Underlying>("</u>"));
+
+  out.AddRule("strikethrough", str_to_underlying<CharT, Underlying>("<s>"),
+              str_to_underlying<CharT, Underlying>("</s>"));
+
   out.AddRule(
       "color",
       [](const TextFormat& f) -> Underlying {
```
Comment (caden-james) [inline Group-04/lib/SerializerHTML.hpp:44]: You could've even used a template lambda here to make your code my flexible.

File: Group-04/app/RichTextState.hpp [hunk 18]
```diff
@@ -0,0 +1,249 @@
...
+    if (start >= m_text.size() || start + count > m_text.size()) return "";
+    m_edit = {start, start + count};
+    return m_text.to_string().substr(start, count);
+  }
+
+  void edit_change(std::string replace) {
+    // Save current state before making changes
+    save_state();
+
+    if (m_edit.first < m_text.size())
```
Comment (caden-james) [inline Group-04/app/RichTextState.hpp:82]: To be more robust and safe you could take the minimum of the m_edit.second and the m_text.size() and use that instead of m_text.size().

File: Group-04/lib/SerializerRTF.hpp [hunk 19]
```diff
@@ -0,0 +1,62 @@
...
+namespace cse {
+
+// RTF serializer template for RichText
+template <typename CharT = char, typename Underlying = std::basic_string<CharT>>
+  requires std::derived_from<Underlying, std::basic_string<CharT>>
+auto SerializerRTF() {
+  struct BasicRichText<CharT, Underlying>::Serializer out{"RTF"};
+
+  // RTF document structure
+  out.header = str_to_underlying<CharT, Underlying>(
```
Comment (caden-james) [inline Group-04/lib/SerializerRTF.hpp:23]: Not sure at the moment how it could be easier to do this but I feel like whenever I see repeating code like this there's always a way to do this without a bunch of repeated statements.

**Review by caden-james — COMMENTED** [review]
Really great and robust application built here. It was really hard to find issues and most of my comments are just on better bounds checking, more modern C++ practices, and potential more efficient solutions. I would focus on making your code more reusable in other classes by obfuscating similar code structures into helper functions. Regardless great work making a responsive web app, it's pretty fun to use too and I could definitely see myself using it especially for Markdown considering I always get confused with it lol. I'll give some general comments on my areas for review:

Implementation Quality:
- I would look to see where bounds safety is viable. I would just make sure any time you're checking for the start and end of your container you could just have a helper function that checks for any potential issues with those bounds.
- I would do the same for memory safety considering the use of Emscripten objects. Make sure they're cleaned up in your web app properly, although I know it's tricky considering the conversion between JS and C++.
- I left a few comments on pre-allocations for strings during serialization. This could provide some minor performance issues.
- You could also implement some way to propagate errors between C++ and JS.

Modern C++: 
- I see some areas where concepts could be used. I'm not the most comfortable using them so I could be overthinking it but especially for the serializer requirements I see some potential use cases. I left in-line comments for that too.
- Same goes for constexpr functions, I left comments where I saw the attribute be potentially viable.
- Some functions could be more efficient and different C++ structures could be replaced with more efficient ones. Again I just left some comments for this.

Again great work! Really impressed here.

File: Group-04/app/web/index.js (hunk 1 above)
Comment (132ikl) [reply Group-04/app/web/index.js:198]: It's supposed to be like the "apply color" button in word processors where you can click on the button to re-apply the already selected color. 

image

It is a bit confusing in the current implementation however, so maybe we should make selecting a color not apply the color and only apply it when the button is pressed.

File: Group-04/app/RichTextState.hpp (hunk 8 above)
Comment (132ikl) [reply Group-04/app/RichTextState.hpp:243]: We only push to/pop from the back, not the front, so I don't think a deque would have any speedup here

File: Group-04/app/RichTextState.hpp (hunk 10 above)
Comment (132ikl) [reply Group-04/app/RichTextState.hpp:212]: The `apply_format` from `RichText` already has boundary checks, so no need to do them for each usage of `apply_format` :smile:

File: Group-04/lib/BitVector.hpp (hunk 11 above)
Comment (132ikl) [reply Group-04/lib/BitVector.hpp:50]: you mean like, `~BitVector()`? That doesn't return a value so `[[nodiscard]]` isn't applicable, unless you mean something else?

File: Group-04/lib/CseAssert.hpp (hunk 16 above)
Comment (132ikl) [reply Group-04/lib/CseAssert.hpp:94]: Since the assert uses `std::cerr` specifically, I think it makes sense to use that in the concept. Maybe it should have a better name, but I don't have anything off the top of my head (`is_printable_to_stderr` doesn't quite do it for me :sweat_smile:)

**Comment by KrishPatel198** [comment]
> ## Bug
> * (-) Multiple color changes occasionally break some text. A significant bug occurs when attempting to change colors. Specifically, after setting “Hello” to orange and “world!” to blue, and then trying to change both to black, the colors fail to update. Additionally, the text becomes uneditable. This issue disrupts the app's flow and needs to be addressed. A possible fix could be ensuring that color changes reset correctly and that text remains editable.
> * (-) Custom not updating in the HTML file type view (as seen in demo).
>

This Issue has been fixed now, just pushed the change for it!

File: Group-04/lib/BitVector.hpp (hunk 12 above)
Comment (CircleShift) [reply Group-04/lib/BitVector.hpp:48]: Even if the pattern is zero, zero is still a valid pattern (although a trivial one since it's just clearing the BitVector)

File: Group-04/lib/BitVector.hpp (hunk 11 above)
Comment (CircleShift) [reply Group-04/lib/BitVector.hpp:50]: This is admittedly a good idea, though it's a little above my comfort level

File: Group-04/lib/SerializerHTML.hpp (hunk 17 above)
Comment (CircleShift) [reply Group-04/lib/SerializerHTML.hpp:44]: Since we're already inside a template function with Underlying as a template class, I think it's ok
